### PR TITLE
Api controllers

### PR DESCRIPTION
**Pull request by TheAlexFreeman**
#Latest version of API project

## change proposal
Additions:
API models
API controllers
API tests
Comments on other tests

**added**
aspnet/Xyz.Provider.Api/ApiModelFactory.cs 
aspnet/Xyz.Provider.Api/Controllers/AddressController.cs 
aspnet/Xyz.Provider.Api/Controllers/ComplexController.cs 
aspnet/Xyz.Provider.Api/Controllers/GenderController.cs 
aspnet/Xyz.Provider.Api/Controllers/ProviderController.cs 
aspnet/Xyz.Provider.Api/Controllers/RoomController.cs 
aspnet/Xyz.Provider.Api/Mapper.cs 
aspnet/Xyz.Provider.Api/Models/ApiAddress.cs 
aspnet/Xyz.Provider.Api/Models/ApiAmenity.cs 
aspnet/Xyz.Provider.Api/Models/ApiComplex.cs 
aspnet/Xyz.Provider.Api/Models/ApiGender.cs 
aspnet/Xyz.Provider.Api/Models/ApiNotifications.cs 
aspnet/Xyz.Provider.Api/Models/ApiProvider.cs 
aspnet/Xyz.Provider.Api/Models/ApiRoom.cs 
aspnet/Xyz.Provider.Api/Models/ApiRoomType.cs 
aspnet/Xyz.Provider.Api/Models/ApiTrainingCenter.cs 

**Many files changed due to comments.** 
## reviewer

- @escalonn 

## contributing

this project is maintained under the terms of the [Contribution Guidelines][contribution-guidelines-url].

[contribution-guidelines-url]: https://github.com/revaturexyz/providerxyz/blob/master/.github/CONTRIBUTING.md 'CONTRIBUTION GUIDELINES'

### CODE REVIEW

File: aspnet/Xyz.Provider.Api/Startup.cs [hunk 1]
```diff
@@ -28,16 +28,19 @@ public void ConfigureServices(IServiceCollection services)
...
+        options.AddPolicy("RevatureCorsPolicy",
             builder =>
-            builder.WithOrigins("http://localhost:4200"
-                              , "https://localhost:4200"
-                              , "https://housingpre.revature.xyz"
-                              , "http://housingpre.revature.xyz")
+            builder.WithOrigins("http://localhost:4200",
+                              "https://localhost:4200",
+                              "https://housing.revature.xyz",
+                              "http://housing.revature.xyz")
```
Comment (escalonn) [inline aspnet/Xyz.Provider.Api/Startup.cs]: we need `housingpre` in here as well as `housing`.

File: aspnet/Xyz.Provider.Api/appsettings.json [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
+    "Instance": "https://identityxyz.b2clogin.com",
+    "ClientId": "fac95bb7-bd11-4ecf-b30b-fdd6ddd13bcd",
+    "Policy": "b2c_1_signup_signin",
+    "Tenant": "identity.revature.xyz",
+    "ScopeRead": "read_any"
+  },
+  
+  "ConnectionStrings": {
+    "DefaultConnection": "data source=.\\SQLEXPRESS;initial catalog=RevHousingDB;integrated security=True;MultipleActiveResultSets=true;"
+  },
```
Comment (escalonn) [inline aspnet/Xyz.Provider.Api/appsettings.json]: this string doesn't belong in here

**Comment by escalonn** [comment]
there are plenty of endpoints (like all of GenderController) that i don't see any need for - YAGNI principle
but it can go in broker for now

File: aspnet/Xyz.Provider.Api/Startup.cs (hunk 1 above)
Comment (TheAlexFreeman) [reply aspnet/Xyz.Provider.Api/Startup.cs]:
```suggestion
                              "http://housingpre.revature.xyz")
```

File: aspnet/Xyz.Provider.Api/Startup.cs [hunk 3]
```diff
@@ -28,16 +28,19 @@ public void ConfigureServices(IServiceCollection services)
...
-        options.AddPolicy("CorsPolicy",
+        options.AddPolicy("RevatureCorsPolicy",
             builder =>
-            builder.WithOrigins("http://localhost:4200"
-                              , "https://localhost:4200"
-                              , "https://housingpre.revature.xyz"
-                              , "http://housingpre.revature.xyz")
+            builder.WithOrigins("http://localhost:4200",
+                              "https://localhost:4200",
+                              "https://housing.revature.xyz",
```
Comment (TheAlexFreeman) [inline aspnet/Xyz.Provider.Api/Startup.cs]:
```suggestion
                              "https://housingpre.revature.xyz",
```

File: aspnet/Xyz.Provider.Api/Properties/launchSettings.json [hunk 4]
```diff
@@ -3,9 +3,7 @@
     "Xyz.Provider.Api": {
       "commandName": "Project",
       "launchBrowser": true,
-      "environmentVariables": {
-        "ASPNETCORE_ENVIRONMENT": "Development"
-      },
+      "environmentVariables": {},
       "applicationUrl": "https://localhost:5001;http://localhost:5000"
     }
   }
```
Comment (escalonn) [inline aspnet/Xyz.Provider.Api/Properties/launchSettings.json]: this file's been changed around a bunch removing the normal visual studio profile and such in previous merges to broker - i'm not sure why

File: aspnet/Xyz.Provider.DataAccess/Entities/RevatureHousingDbContext.cs [hunk 5]
```diff
@@ -14,6 +14,7 @@ public RevatureHousingDbContext()
     public RevatureHousingDbContext(DbContextOptions<RevatureHousingDbContext> options)
         : base(options)
     {
+      this.Database.EnsureCreated();
```
Comment (escalonn) [inline aspnet/Xyz.Provider.DataAccess/Entities/RevatureHousingDbContext.cs]: no... i've already said migrations and ensurecreated are mutually exclusive.

File: aspnet/Xyz.Provider.DataAccess/Repository/AddressRepository.cs [hunk 6]
```diff
@@ -18,10 +18,11 @@ public AddressRepository(RevatureHousingDbContext dbContext, ILogger<IComplexRep
     {
       _dbContext = dbContext ?? throw new ArgumentNullException(nameof(dbContext));
       _logger = logger;
+
     }
 
     /// <summary>
-    /// Asynchronously adds address to database
+    /// Asynchronusly adds address to database
```
Comment (escalonn) [inline aspnet/Xyz.Provider.DataAccess/Repository/AddressRepository.cs]: why would you intentionally make it misspelled again

File: aspnet/Xyz.Provider.DataAccess/Repository/GenderRepository.cs [hunk 7]
```diff
@@ -25,10 +23,10 @@ public GenderRepository(RevatureHousingDbContext dbContext,ILogger<IGenderReposi
     /// <summary>
     /// NOT IMPLEMENTED - Asynchronusly removes a gender from database
     /// </summary>
-    /// <param name="Id"></param>
+    /// <param name="ID"></param>
     /// <param name="ProviderId"></param>
     /// <returns></returns>
-    public Task RemoveAsync(int Id, int ProviderId)
+    public Task RemoveAsync(int ID, int ProviderId)
```
Comment (escalonn) [inline aspnet/Xyz.Provider.DataAccess/Repository/GenderRepository.cs]: why would you make this change?
both are incorrect, it should be `id` and `providerId`.
yes, in the strings it should be "ID".
but really, this method shouldn't even exist

File: aspnet/Xyz.Provider.Lib/Interface/IRepository.cs [hunk 8]
```diff
@@ -9,11 +9,11 @@ namespace Xyz.Provider.Lib.Interface
...
     //IEntity are the Library Models
     //The other interfaces extend this interface which
-    //  implements the generic methods that are to access the DBContext
+    //  impliments the generic methods that are to access the DBContext
   {
     Task<IEntity> GetAsync(int Id);
     Task<IEntity> AddAsync(IEntity entity, int ProviderId);
     Task UpdateAsync(IEntity entity, int ProviderId);
-    Task RemoveAsync(int Id, int ProviderId);
+    Task RemoveAsync(int ID, int ProviderId);
```
Comment (escalonn) [inline aspnet/Xyz.Provider.Lib/Interface/IRepository.cs]: sloppy style!
all these parameters should be `id` and `providerId`

File: aspnet/Xyz.Provider.Lib/Interface/IRepository.cs [hunk 9]
```diff
@@ -9,11 +9,11 @@ namespace Xyz.Provider.Lib.Interface
   public interface IRepository<IEntity> where IEntity : class
     //IEntity are the Library Models
     //The other interfaces extend this interface which
-    //  implements the generic methods that are to access the DBContext
+    //  impliments the generic methods that are to access the DBContext
```
Comment (escalonn) [inline aspnet/Xyz.Provider.Lib/Interface/IRepository.cs]: why would you add a misspelling?

**Pull request merged**